examples/advanced/visual_sudoku.py

```diff
@@ -0,0 +1,281 @@
+    sudoku_torch_dimension = sudoku_puzzle.shape + (N_COLORS, IMAGE_WIDTH, IMAGE_HEIGHT,)
+
+    vizsudoku = torch.zeros(sudoku_torch_dimension, dtype=torch.float32)
+
+    # TODO: explain standard transformation?
+    transform = transforms.Compose([transforms.ToTensor(),
+                                    transforms.Normalize((0.5,), (0.5,))])
+
+    # Download and load the MNIST data
+    testset = datasets.MNIST('.', download=True, train=False, transform=transform)
+
+    # Dictionary (number, corresponding sample images)
+    digit_indices = {number: torch.LongTensor(*np.where(testset.targets == number)) for number in range(1,10)}
+
+    # replace all numeric values in sudoku torch tensor grid by an mnist digit
+    
+    # TODO: Why do you not use np.ndenumerate ? and then for each number in the 
+    #       grid you loop select a random image?
+    # IS it for having differnet images ? if yes, motivate or comment :)
+    # TODO: PROPOSED:
+    for idx, val in np.ndenumerate(sudoku_puzzle[sudoku_puzzle > 0]):
+        image_tensor = np.random.choice(digit_indices[val])
+        vizsudoku[idx] = image_tensor
+
+    ## OLD:
+    for val in np.unique(sudoku_puzzle[sudoku_puzzle > 0]):
+
+        val_idx = np.where(sudoku_puzzle == val)
+        idx = torch.LongTensor(np.random.choice(digit_indices[val], len(sudoku_puzzle[val_idx])))
+        vizsudoku[val_idx] = torch.stack([testset[i][0] for i in idx])
+
+    return vizsudoku
+
+
+
+from torch import nn 
+import torch.nn.functional as F
+
+class LeNet(nn.Module):
+    '''
+        TODO: Write doc+ reference to paper
+    '''
+    def __init__(self):
+        super(LeNet, self).__init__()
+        self.conv1 = nn.Conv2d(1, 6, 5, 1, padding=2)
+        self.conv2 = nn.Conv2d(6, 16, 5, 1)
+        self.fc1 = nn.Linear(5*5*16, 120)
+        self.fc2 = nn.Linear(120, 84)
+        self.fc3 = nn.Linear(84,10)
+
+    def forward(self, x):
+        x = F.relu(self.conv1(x))
+        x = F.max_pool2d(x, 2, 2)
+        x = F.relu(self.conv2(x))
+        x = F.max_pool2d(x, 2, 2)
+        x = x.view(-1, 5*5*16) 
+        x = F.relu(self.fc1(x))
+        x = self.fc2(x)
+        x = self.fc3(x)
+        return F.log_softmax(x, dim=1)
+
+def load_clf(clf_classname, path):
+    '''
+        TODO: Write doc.
+    '''
+    net = clf_classname()
+    state_dict = torch.load(path, map_location=lambda storage, loc: storage)
+    net.load_state_dict(state_dict)
+    return net
+
+
+@torch.no_grad() #TODO: What does this do ?
+def predict_proba_sudoku(model, vizsudoku):
+    #TODO: Change shape of output to work for differnet kinds of sudoku shapes (4x4, 9x9, ...)
+    # reshape from 9x9x1x28x28 to 81x1x28x28
+    pred = model(vizsudoku.flatten(0,1))
+    # our NN returns 81 probabilistic vector: an 81x10 matrix
+    return pred.reshape(9,9,10).detach() # reshape as 9x9x10 tensor for easier visualisation
+
+if __name__ == '__main__':
```

Comment:
create a function `def visual_sudoku()` that you call here (rest is blank).

Put that `def visual_sudoku()` at the top: make your script readable from top to bottom